### ISSUE TITLE
Need LDM decisions for overload resolution in the face of 'out var'

### ISSUE THREAD

**Issue by gafter** [dotnet/roslyn#13845]
We need the LDM to decide how 'out var' should interact with overload resolution. Currently, the presence of an out var argument prevents that parameter position from contributing to betterness. For example, the following program is ambiguous in the current prototype.

I would recommend saying that there is an implicit "conversion from expression" from the 'out var' to any (`out`) argument type. Then, the usual tie-breakers will cause the first of these methods to be selected.

```cs
class Program
{
    static void Main(string[] args)
    {
        M(out var t); // error: ambiguous?
    }
    static void M(out string s)
    {
        s = null;
    }
    static void M(out object o)
    {
        o = null;
    }
}
```

**Comment by gafter** [comment]
The draft spec says

> For the purposes of overload resolution (see sections 7.5.3.2 Better function member and 7.5.3.3 Better conversion from expression), neither conversion is considered better when corresponding argument is an implicitly-typed out variable declaration.

This appears to be intended to override the parts of the current specification that compare only the method parameters, and ignore the argument and its conversions.

**Comment by gafter** [comment]
Based on the ECMA spec, I propose

--------------------------------------------------

Overload resolution is modified as follows:

We add a new conversion:

> There is a *conversion from expression* from an implicitly-typed out variable declaration to every type.

The section "Better Conversion from expression" (from the ECMA version, which is the base of our draft C# spec) is modified to add the bold text, below:

> 13.6.4.4 Better conversion from expression
Given an implicit conversion C1 that converts from an expression E to a type T1, and an implicit conversion C2 that converts from an expression E to a type T2, C1 is a better conversion than C2 if at least one of the following holds:
- E has a type S and an identity conversion exists from S to T1 but not from S to T2
- E is not an anonymous function **or implicitly-typed out variable declaration,** and T1 is a better conversion target than T2 (§13.6.4.6)
- E is an anonymous function, T1 is either a delegate type D1 or an expression tree type `Expression<D1>`, T2 is either a delegate type D2 or an expression tree type `Expression<D2>` and one of the following holds:
  - D1 is a better conversion target than D2
  - D1 and D2 have identical parameter lists, and one of the following holds:
    - D1 has a return type Y1, and D2 has a return type Y2, an inferred return type X exists for E in the context of that parameter list (§13.6.3.13), and the conversion from X to Y1 is better than the conversion from X to Y2
    - E is async, D1 has a return type `Task<Y1>`, and D2 has a return type `Task<Y2>`, an inferred return type `Task<X>` exists for E in the context of that parameter list (§13.6.3.13), and the conversion from X to Y1 is better than the conversion from X to Y2
    - D1 has a return type Y, and D2 is `void` returning

The type of an implicitly-typed out variable is the type of the corresponding parameter in the signature of the method.

**Comment by gafter** [comment]
The specification in https://github.com/dotnet/roslyn/blob/master/docs/features/outvar.md has been updated in 28916283bcbd4097bf39b9977e2e549cf085ea64 to align with the intent of the LDM and the shape of the implementation.

**Comment by gafter** [comment]
The draft has been updated in #13999. I checked with @MadsTorgersen and the LDM does not intend to produce any further specification text for quite some time. He is OK with a spec that aligns with the implementation, the shape of the current specification, and the intent of the LDM as previously expressed. So there is nothing more to do here until it is time to integrate this into the full specification.
